### PR TITLE
Remove _model from ModelMeshPartPayload

### PR DESCRIPTION
**Pull request by SamGondelman**
Instead of repeatedly checking properties of _model in ModelMeshPartPayload::getShapeKey(), getKey(), and getLayer(), we only need to update those properties when they could have changed, which we can do in updateRenderItems().  Also some other small Model-related cleanup.

Although not extremely expensive, these functions are called very often.  Here, for example, you can see the effect of moving the computation to setShapeKey() and calling it only when things have changed, in the same scene:
Before:
getshapekey1
Now:
getshapekey2

Test plan:
- Go to an empty domain.  Reload content.  Clear your KTX cache (Developer -> Render -> Clear KTX Cache).  Restart.
- Go to a domain with lots of model entities, like dev-welcome or AvatarIsland (or SamsWorld/leviathan if AvatarIsland is on a different protocol version).  Verify that all models load and that if you wait, all of their textures load.  Bring in some models from the marketplace.  Verify that they appear correctly.
- Repeatedly toggle the visibility on a model using Create.  Verify that the model's visibility always matches the checkbox.
- Manually edit the textures on a model (select a model with textures, replace some of the links to random pictures).  Verify that the textures on the model actually change when you replace the links.
- Replace a model's URL with a different model URL.  Verify that the model actually switches to use the new URL.
- Make a model with an invalid URL.  You should see a green cube.  Replace the URL with a valid URL and verify that the model appears and the green cube disappears.
- Have someone else join you.  Make sure their avatar renders.  Have them change avatars.  You should see the change.
- Make sure you can see your own avatar, and that it updates when you change avatars.
- Switch between first and third person.  You should be able to see your head in third person.  When in first person, if you walk backwards, you should not see your eyes/mouth.  In HMD, when you move your head around, you should not see your eyes/mouth.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5745/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5745/HighFidelity-Beta-PR12026-83fc6abc241c921fcbd63a8fe45728f75590ee4f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5745/HighFidelity-Beta-PR12026-83fc6abc241c921fcbd63a8fe45728f75590ee4f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5754/app_12026_PR.apk

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 1]
```diff
@@ -388,94 +386,65 @@ void ModelMeshPartPayload::updateTransformForSkinnedMesh(const Transform& render
...
 
-int ModelMeshPartPayload::getLayer() const {
-    ModelPointer model = _model.lock();
-    if (model) {
-        if (model->isLayeredInFront()) {
-            return Item::LAYER_3D_FRONT;
-        } else if (model->isLayeredInHUD()) {
-            return Item::LAYER_3D_HUD;
-        }
+    if (isLayeredInFront || isLayeredInHUD) {
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.cpp]: Maybe should just combine these as one argument to the function?
```
bool isLayered = isLayeredInFront || isLayeredInHUD;
evalKey(isVisible, isLayered);
```

File: libraries/render-utils/src/MeshPartPayload.cpp [hunk 2]
```diff
@@ -388,94 +386,65 @@ void ModelMeshPartPayload::updateTransformForSkinnedMesh(const Transform& render
...
-        return ShapeKey::Builder::invalid(); // FIXME
+int ModelMeshPartPayload::evalLayer(bool isLayeredInFront, bool isLayeredInHUD) const {
+    if (isLayeredInFront) {
+        return Item::LAYER_3D_FRONT;
+    } else if (isLayeredInHUD) {
+        return Item::LAYER_3D_HUD;
+    } else {
+        return Item::LAYER_3D;
     }
+}
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.cpp]: `ModelMeshPartPayload::evalLayer()` does not depend on any `ModelMeshPartPayload` interal data members --> it could be `static`, or it could be just a lookup table or `switch` statement.

File: libraries/render-utils/src/MeshPartPayload.h [hunk 3]
```diff
@@ -112,16 +117,14 @@ class ModelMeshPartPayload : public MeshPartPayload {
...
 
-    enum State : uint8_t {
-        WAITING_TO_START = 0,
-        STARTED = 1,
-    };
-
-    mutable State _state { WAITING_TO_START } ;
+    render::ItemKey _itemKey { render::ItemKey::Builder::opaqueShape().build() };
+    int _layer { render::Item::LAYER_3D };
+    render::ShapeKey _shapeKey { render::ShapeKey::Builder::invalid() };
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/MeshPartPayload.h]: Not a blocker but:  [the Lost Art of C Structure Packing](http://www.catb.org/esr/structure-packing/) suggests you put `int _layer` at the bottom of this list.

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -237,20 +255,32 @@ void Model::updateRenderItems() {
...
+            transaction.updateItem<ModelMeshPartPayload>(itemID, [modelTransform, clusterMatrices, invalidatePayloadShapeKey,
+                    isWireframe, isVisible, isLayeredInFront, isLayeredInHUD](ModelMeshPartPayload& data) {
                 data.updateClusterBuffer(clusterMatrices);
                 Transform renderTransform = modelTransform;
                 if (clusterMatrices.size() == 1) {
                     renderTransform = modelTransform.worldTransform(Transform(clusterMatrices[0]));
                 }
                 data.updateTransformForSkinnedMesh(renderTransform, modelTransform);
+
+                data.setKey(data.evalKey(isVisible, isLayeredInFront, isLayeredInHUD));
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: AFAICT: `data.evalKey()` is only ever called as an argument to `data.setKey()`.

This means you could combine the two calls into one:  `data.setKey(isVisible, isLayeredInFront, isLayeredInHUD);`]

Similarly for `evalLayer()`.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5763/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5763/HighFidelity-Beta-PR12026-2ba3e337db8a122d9ecfb49c7cec853f1f257547.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5763/HighFidelity-Beta-PR12026-2ba3e337db8a122d9ecfb49c7cec853f1f257547.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5781/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5782/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5786/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5786/HighFidelity-Beta-PR12026-69c1c65eb8853a6ebd0701c53f6b732d962c138a.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5792/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5792/HighFidelity-Beta-PR12026-69c1c65eb8853a6ebd0701c53f6b732d962c138a.dmg

**Comment by SamGondelman** [comment]
build this PLEASE

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5801/app_12026_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5801/HighFidelity-Beta-PR12026-69c1c65eb8853a6ebd0701c53f6b732d962c138a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5801/HighFidelity-Beta-PR12026-69c1c65eb8853a6ebd0701c53f6b732d962c138a.exe

File: libraries/render-utils/src/Model.h [hunk 5]
```diff
@@ -346,11 +345,7 @@ public slots:
     // hook for derived classes to be notified when setUrl invalidates the current model.
     virtual void onInvalidate() {};
 
-
-protected:
```
Comment (samcake) [inline libraries/render-utils/src/Model.h]: WHy did you remove this ?

Comment (SamGondelman) [reply libraries/render-utils/src/Model.h]: There's already a protected section right above this, so it was unnecessary.

**Comment by cain-kilgore** [comment]
Testing.

**Comment by cain-kilgore** [comment]
# Test Passed
Test | Result
--- | ---
Go to an empty domain. Reload content. Clear your KTX cache (Developer -> Render -> Clear KTX Cache). Restart. | 👍 
Go to a domain with lots of model entities, like dev-welcome or AvatarIsland (or SamsWorld/leviathan if AvatarIsland is on a different protocol version). Verify that all models load and that if you wait, all of their textures load. Bring in some models from the marketplace. Verify that they appear correctly. | 👍 
Repeatedly toggle the visibility on a model using Create. Verify that the model's visibility always matches the checkbox. | 👍 
Manually edit the textures on a model (select a model with textures, replace some of the links to random pictures). Verify that the textures on the model actually change when you replace the links. | 👍 
Replace a model's URL with a different model URL. Verify that the model actually switches to use the new URL. | 👍 
Make a model with an invalid URL. You should see a green cube. Replace the URL with a valid URL and verify that the model appears and the green cube disappears. | 👍 
Have someone else join you. Make sure their avatar renders. Have them change avatars. You should see the change. | 👍 
Make sure you can see your own avatar, and that it updates when you change avatars. | 👍 
Switch between first and third person. You should be able to see your head in third person. When in first person, if you walk backwards, you should not see your eyes/mouth. In HMD, when you move your head around, you should not see your eyes/mouth. | 👍  (See notes)

When jumping in first person, you can see the back of your head for a split second and it looks somewhat strange. This has persisted in master for quite some time and might be worth at _some point_ fixing.
